### DISCUSSION TITLE
Command with root permission

### DISCUSSION THREAD

**Discussion by noahfraiture** [Aylur/ags#608] [Q&A] [answered]
Hello,
I'm on nixos and I'm trying to make a widget to switch my specialisation, here's a small part of my code : 
```
Widget.Button({
    vpack: "center",
    class_name: "random-wallpaper",
    label: "light",
    on_primary_click: () => {
      Utils.exec(
        `sudo /nix/var/nix/profiles/system/specialisation/light/bin/switch-to-configuration switch`,
      );
      Utils.exec(
        `nu -c "ps | find hx | each { kill -s 10 $in.pid }" ; nu -c "ps | find ags | each { kill $in.pid }"; hyprpanel`,
      );
    },
  })
```
But I have two problem, the first one is that to use `switch-to-configuration`, I need sudo access, and ags doesn't have them. I call the widget with a hyprland binding and thus the command `ags -t widget` and it's impossible to put a sudo there due to hyprland system.
My second problem is that the call `hyprpanel` which is a wrapper for ags widget, is attached to the environment and thus the widget freeze because of non returning call.
Do you know how I can solve both issue ?

**Comment by Aylur** [comment, marked as answer]
1. maybe you could use polkit? `pkexec`
2. don't use `exec`, it will block the eventloop until it returns, use `execAsync`

**Reply by noahfraiture** [reply]
Thank you, smart for pkexec. But I have an issue with it unrelated to ags. `The value for the SHELL variable was not found in the /etc/shells file`. I defined my shell for my user and for default user. I use hyprland polkit for my polkit. Is there something you can think of that I'm missing ?

**Reply by Aylur** [reply]
try
```js
Utils.execAsync([
  "bash", "-c", "SHELL=/run/current-system/sw/bin/bash pkexec path-to-script"
])
```

**Reply by noahfraiture** [reply]
Oh great it works thank you !
One last thing related to ags this time. I can't kill hyprpanel/ags from the widget. I don't know if it's because ags is trying to kill hyprpanel which is basically a suicide, or it's because it's called by a different user.
In this 
```
  const switchTheme = (theme: string) => {
    Utils.exec([
      "bash",
      "-c",
      `SHELL=/run/current-system/sw/bin/bash pkexec /nix/var/nix/profiles/system/specialisation/${theme}/bin/switch-to-configuration switch`,
    ]);
    Utils.exec([
      "nu",
      "-c",
      "ps | find hx | each { kill -s 10 $in.pid }",
    ]);
    Utils.exec("hyprpanel -q");
    Utils.execAsync("hyprpanel");
  };
```
The call `hyprpanel -q` doesn't work. However, if I try to do it in a cli as a regular user, it works. 
Is there a specific way to reload ags widgets ? Hyprpanel doesn't have a special way of doing it, it seems

**Reply by Jas-SinghFSU** [reply]
Why not run hyprpanel on a different bus and kill that with `ags -q -b busName` or `hyprpanel -q -b busName`?

**Reply by noahfraiture** [reply]
Oh yeah thank you. It wasn't working at the beginning because of bad configuration at another place but this solution actually works !
